# A Course in Python
## Chapter 3 - Problems

In [1]:
import math
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt
import nltk

# 1) 
Find the number of positive integers $ 0 < n < 20000 $ such that 1997 divides $n^2 + (n+1)^2 $.  Try the same code for 2009 and 2022

In [13]:
dividesBy1997 = []
dividesBy2009 = []
dividesBy2022 = []

for n in range(1, 20001): 
    x = n**2 + (n + 1)**2
    if(x % 1997 == 0):
        dividesBy1997.append(n)
    if(x % 2009 == 0):
        dividesBy2009.append(n)
    if(x % 2022 == 0):
        dividesBy2022.append(n)

print(f"dividesBy1997: {dividesBy1997}")
print(f"dividesBy2009: {dividesBy2009}")
print(f"dividesBy2022: {dividesBy2022}")

dividesBy1997: [792, 1204, 2789, 3201, 4786, 5198, 6783, 7195, 8780, 9192, 10777, 11189, 12774, 13186, 14771, 15183, 16768, 17180, 18765, 19177]
dividesBy2009: []
dividesBy2022: []


# 2) 
Show that the number of $k$ between 0 and 1000 for which $\binom{1000}{k}$ is odd is a power of 2.  
Note that $\binom{n}{m}$ is the binomial coefficient defined by <br/>
$$\binom{n}{m}=\frac{n!}{m!(n-m)!}$$  
and it is available in the python `math` library as the command `comb(n,m)`

In [79]:
import math

theNumberOfKBetween0And1000 = []
# between 0 and 1000
for k in range(0, 1001):
    binomOutput = math.comb(1000, k)
    rhsOutput = math.factorial(1000) / (math.factorial(k) * math.factorial(1000 - k))
    
    if(binomOutput % 2 != 0):
        if( math.sqrt(binomOutput).is_integer() ):
            theNumberOfKBetween0And1000.append(k)
            #print(f"math.comb({k}) result is odd and a power of 2")

print(f"These are the inputs for the math.comb(k) outputs that are odd and a power of 2; \n{theNumberOfKBetween0And1000}")

These are the inputs for the math.comb(k) outputs that are odd and a power of 2; 
[0, 32, 40, 64, 72, 96, 104, 128, 136, 160, 168, 192, 200, 224, 232, 256, 264, 288, 296, 320, 328, 352, 360, 384, 392, 416, 424, 448, 456, 480, 488, 512, 520, 544, 552, 576, 584, 608, 616, 640, 648, 672, 680, 704, 712, 736, 744, 768, 776, 800, 808, 832, 840, 864, 872, 896, 904, 928, 936, 960, 968, 1000]


# 3) 
For integers $2 \leq n \leq 200$, find all $n$ such that $n$ divides $(n-1)! + 1$.  
Show that there are 46 such  $n$.

In [85]:
import math

count = 0
for n in range(2, 201):
    if( (n / (math.factorial(n - 1) + 1)).is_integer() ):
        count += 1

if(count == 46):
    print("There are 46 such n that divides (n-1)!+1")
else:
    print(f"There are not 46 such n that divides (n-1)!+1. \nOnly {count} were discovered")


There are not 46 such n that divides (n-1)!+1. 
Only 23 were discovered


# 4) 
Show that there is only one positive integer $n$ smaller than 100 such that $n! + (n+1)!$ is the square of an integer

In [95]:
count = 0;
for n in range(1, 100):
    result = math.factorial(n) + math.factorial(n + 1)
    if( math.sqrt(result).is_integer() ):
        count += 1;

print(f"There are {count} positive input integers")

There are 73 positive input integers


# 5)
Let $m$ be a natural number and
$$
A=\frac{(m+3)^3+1}{3 m}
  $$

Find all the integers $m$ less than 500 such that $A$ is an integer. Show that $A$ is always odd.

In [18]:

allValues = []

for m in range(1, 500):
    value = ((m + 3)**3 + 1) / (3 * m)
    if( value.is_integer() ):
        allValues.append(value)
        
allValues

[21.0, 117.0]

# 6)
For a given 𝑛, calculate the series
$$ 1^3 + 2^3 + ... + n^3, 1 + \frac{1}{1} + \frac{1}{2!} + ... + \frac{1}{n!} $$ 


In [21]:
import math
n = 5

values = []
x, y = 0, 0
for n in range(1, n+1):
    x += n**3
    y += round(1 / math.factorial(n), 3)

print(f"Sum of cubes (S1) for n={n}: {x}")
print(f"Sum of reciprocal factorials (S2) for n={n}: {y}")

Sum of cubes (S1) for n=5: 225
Sum of reciprocal factorials (S2) for n=5: 1.717


# 7)
Show that only $n$ less than 1000 such that
$$ 3^n + 4^n + ... +  (n + 2)^n = (n + 3)^n $$
are the numbers 2 and 3

In [37]:
import math

n = 32
lhs = 0
for n in range(n):
    number = (n+1) + 2
    lhs += ((n+1) + 2)**n  

rhs = (n + 3)**n

print(f"lhs = {lhs}")
print(f"rhs = {rhs}")

lhs = 302758531707251096758465027961350381747822046796
rhs = 299120672332806228664106719451209941853702979584


# 8)
find all the numbers up to one million which have the following property:  
if $ n = d_1 d_2 ... d_k $ then $ n = d_1! + d_2! + ... + d_k! $ (eg 145 = 1! + 4! + 5!) 

# 9) 
Consider the number 485. Observe that 485 + 584 = 1069 is a prime number.
Find all numbers 𝑛 between 1 and 1000 such that 𝑛 plus its reverse is prime.

In [2]:
x = 486
x_r = int(str(x)[::-1])
x_r

684

# 10) 
write two functions $f$ and $g$, each accepting two sequences $(x_1, x_2, ..., x_n)$ and $(y_1, y_2, ...,y_n)$ of positive integers, and return respectively.  }
$$

$$
\sqrt{x_1 y_1} + \sqrt{x_2, y_2} \cdots + \sqrt{x_n y_n}
$$

$$
\sqrt{x_1+x_2+\cdots+x_n} \times \sqrt{y_1+y_2+\cdots+y_n}
$$

# 11) 
Write a function that steps in sequence $(a_1, a_2, ..., a_n)$ and returns $\frac{(a_1 + a_2 + ... + a_n)}{n}$  
Use this function to demonstrate Chebycheff's inequality ,  
i.e
$$
\frac{a_1+a_2+\cdots+a_n}{n} \times \frac{b_1+b_2+\cdots+b_n}{n} \leq \frac{a_1 b_1+a_2 b_2+\cdots+a_n b_n}{n},
$$

for non-increasing sequences of numbers

# 12) 
A number is called a Harshad number if it is divisible by the sum of its digits (e.g.,
12 is Harshad as it is divisible by 1 + 2 = 3). Find all 2-digit Harshad numbers.
How many 5-digit Harshad numbers are there? (Harshad means “giving joy” in
Sanskrit, defined and named by the Indian mathematician D. Kaprekar.)

# 13) 
The formula 𝑒41 = 𝑛2+𝑛+41 produces prime numbers for 0 ≤ 𝑛 ≤ 39 but not for
𝑛 = 40. Check that for no 𝑖 between 1 and 10000 does the formula 𝑒𝑖 = 𝑛2 + 𝑛 +𝑖
produce prime numbers for a larger interval starting at 𝑛 = 0

# 14) 
Recall that Fibonacci numbers are defined recursively by 𝑎1 = 𝑎2 = 1 and
𝑎𝑛 = 𝑎𝑛−1 + 𝑎𝑛−2, for 𝑛 ≥ 2. Show that the following identities hold for 𝑛 = 100.

$$
a_1 + a_2 + ... + a_n = a_{n+2} - 1
$$

$$
a_n^4-a_{n-2} \cdot a_{n-1} \cdot a_{n+1} \cdot a_{n+2}=1
$$

$$
a_{2n} = a_n(a_{n-1}+a_{n+1})
$$

# 15) Compute


$$
\sqrt{1+\frac{1}{1^{2}}+\frac{1}{2^{2}}}+\sqrt{1+\frac{1}{2^{2}}+\frac{1}{3^{2}}}+\cdots+\sqrt{1+\frac{1}{2022^{2}}+\frac{1}{2023^{2}}}
$$

# 16) 
Investigate whether the series


$$
2^{-\frac{1}{2}}+(3+5)^{-\frac{1}{2}}+(7+11+13)^{-\frac{1}{2}}+(17+19+23+29)^{-\frac{1}{2}}+\cdots
$$


converges.

# 17
Write


$$
\prod_{n=1}^{10} \frac{(2 n-1)(2 n+1)}{2 n \times 2 n}=\frac{1 \times 3}{2 \times 2} \frac{3 \times 5}{4 \times 4} \frac{5 \times 7}{6 \times 6} \cdots
$$

# 18)
Investigate


$$
\frac{\mathrm{e}}{2}=\left(\frac{2}{1}\right)^{\frac{1}{2}}\left(\frac{2}{3} \frac{4}{3}\right)^{\frac{1}{4}}\left(\frac{4}{5} \frac{6}{5} \frac{6}{7} \frac{8}{7}\right)^{\frac{1}{8}}\left(\frac{8}{9} \frac{10}{9} \frac{10}{11} \frac{12}{11} \frac{12}{13} \frac{14}{13} \frac{14}{15} \frac{16}{15}\right)^{\frac{1}{16}} \ldots
$$

# 19)
Investigate


$$
\sum_{n=0}^{\infty} \frac{(-1)^{n}}{2 n+1} \sum_{k=0}^{2 n} \frac{1}{2 n+4 k+3}=\frac{3 \pi}{8} \log \frac{\sqrt{5}+1}{2}-\frac{\pi}{16} \log 5
$$

# 20)
Find the smallest number expressible as the sum of two cubes in two different ways (hint: the number is less than 3000 ).

In [107]:
a = 2j
print(a.imag)
print(a.real)

2.0
0.0
